nemo/core/neural_types/elements.py

```diff
@@ -0,0 +1,191 @@
+# ! /usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = [
+    'ElementType',
+    'VoidType',
+    'ChannelType',
+    'AcousticEncodedRepresentation',
+    'AudioSignal',
+    'SpectrogramType',
+    'MelSpectrogramType',
+    'MFCCSpectrogramType',
+    'LogitsType',
+    'LabelsType',
+    'LossType',
+    'RegressionValuesType',
+    'CategoricalValuesType',
+    'PredictionsType',
+    'LogprobsType',
+    'LengthsType',
+    'EmbeddedTextType',
+    'EncodedRepresentation',
+]
+import abc
+from abc import ABC, abstractmethod
+from typing import Dict, Optional, Tuple
+
+from nemo.core.neural_types.comparison import NeuralTypeComparisonResult
+
+
+class ElementType(ABC):
+    """Abstract class defining semantics of the tensor elements.
+    We are replying on Python for inheritance checking"""
+
+    @abstractmethod
+    def __str__(cls):
+        pass
+
+    @property
+    def type_parameters(self) -> Dict:
+        """Override this property to parametrize your type"""
+        return {}
+
+    @property
+    def fields(self) -> Optional[Tuple]:
```

Comment:
Please comment that property.

This is important, as the user should understand what is the distinction/usage of type_parameters and fields.

Can we come up with names that better explain/distringuish those two entities?

Reply:
added comments